src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs

```diff
@@ -1,20 +1,63 @@
 ﻿using System;
 using System.Diagnostics;
+using System.Collections.Generic;
+using System.Linq;
 using Microsoft.VisualStudio.ComponentModelHost;
 using Microsoft.VisualStudio.Shell;
+using EnvDTE;
+using EnvDTE80;
 using Nuget.PackageIndex.Client.Analyzers;
 
 namespace Nuget.PackageIndex.VisualStudio
 {
     /// <summary>
-    /// TODO remove this hack after RC
+    /// This filter is needed only until we enable Missing Package suggestions for all C# projects.
+    /// For now we limit this feature only to ProjectK scenarios.
     /// </summary>
     internal class ProjectKFilter : IProjectFilter
     {
+        private Dictionary<string, string> FilesCache = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
+        private bool _disposed;
+        private SolutionEvents _solutionEvents;
+        private object _cacheLock = new object();
+        private bool _solutionIsClosing;
+
+        public ProjectKFilter()
+        {
+            var container = ServiceProvider.GlobalProvider.GetService<IComponentModel, SComponentModel>();
+
+            // Get the DTE
+            var dte = ServiceProvider.GlobalProvider.GetService(typeof(DTE)) as DTE;
+            Debug.Assert(dte != null, "Couldn't get the DTE. Crash incoming.");
+
+            var events = (Events2)dte.Events;
+            if (events != null)
+            {
+                _solutionEvents = events.SolutionEvents;
+                Debug.Assert(_solutionEvents != null, "Cannot get SolutionEvents");
+
+                // clear all cache if solution closed.
+                _solutionEvents.BeforeClosing += OnBeforeSolutionClosed;
+                _solutionEvents.Opened += OnSolutionOpened;
+                _solutionEvents.ProjectRemoved += OnProjectRemoved;
+                _solutionEvents.ProjectRenamed += OnProjectRenamed;
+            }
+        }
 
         public bool IsProjectSupported(string filePath)
         {
-            return ThreadHelper.JoinableTaskFactory.Run(async delegate
+            if (string.IsNullOrEmpty(filePath) || _solutionIsClosing)
+            {
+                return false;
+            }
+
+            string projectFullName;
+            if (FilesCache.TryGetValue(filePath, out projectFullName))
```

Comment:
About links .. not sure yet what to do with them. Need to open an issue to track that